[Death Knight] Implement Reanimated Shambler

Khazakdk opened the pull request:
This isn't quite ready to merge but putting this up to get feedback on the implementation to make sure the idea is sound.  

While verifying if necroblast double dips on mastery (its not whitelisted in the mastery spelldata so shouldn't), Taez and I were getting in game numbers that didn't match what we calculated they should be.

Using the formula "(str + wdps * 6) * 1.5 * .97 * mastery * vers" we both got numbers lower than what's reported in game.  With the bulk of the implementation done I'm going to return to checking logs and my math to see what I'm missing.  

It's likely a mistake I'm making with the formula.  Though until I find a smoking gun I feel more comfortable leaving this in draft status in case I need to adjust a multiplier.

Khazakdk committed:
[Death Knight] Implement Reanimated Shambler

Melekus commented:
If the damage comes from the pet, maybe it has a specific AP from owner coefficient that you'll have to figure out. Does the pet show up on WCL?

Melekus commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -2666,6 +2674,106 @@ struct magus_pet_t : public death_knight_pet_t
+    necroblast_t( reanimated_shambler_pet_t* player, util::string_view options_str ) :
+      pet_action_t( player, "necroblast", player -> find_spell( 334851 ) )
+    {
+      parse_options( options_str );
```
necroblast is a proc action, it shouldn't use (or have access to) options

Khazakdk replied:
I thought things executed by the APL use options_str.  Does it not count here because the shambler apl is fixed?

Melekus replied:
Well it's only necessary if you plan to add options to necroblast in the shambler APL.

Melekus commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -2666,6 +2674,106 @@ struct magus_pet_t : public death_knight_pet_t
+    }
+
+    timespan_t execute_time() const override
+    {
+      return timespan_t::from_seconds( const_cast<travel_t*>( this )->rng().gauss( 2.9, 0.2 ) );
+    }
```
Is this the same delay as gargoyle or just WIP?

Khazakdk replied:
WIP.  There seems to be a significant amount of variance on how long it takes to impact after spawning.  I'm thinking both the mean and standard deviation need to go up from what gargoyle but need to do some more testing

Khazakdk replied:
https://docs.google.com/spreadsheets/d/1XJkX7tMaIKpxECNBD9a21PyeDbXmUE4JGxH4y4IkoxU/edit?usp=sharing

Pulled all shambler summon/death events out of a 22 minute log, found the time gap between each summon/death and ran those through wolfram alpha.  Got a mean of 3.449 and sample standard deviation of .2846.  I will round those to 3.5 and 0.28 and update the travel rng in the sim accordingly.

Khazakdk replied:
https://docs.google.com/spreadsheets/d/1XJkX7tMaIKpxECNBD9a21PyeDbXmUE4JGxH4y4IkoxU/edit?usp=sharing

Took a 22 minute log, filtered it down to summon/cast necroblast events and got the delta between each pair of events and ran the values though wolfram alpha.  Got a mean of 4.597 and sample standard deviation of .3399.  I will update the shambler travel rng accordingly.

Khazakdk commented:
> If the damage comes from the pet, maybe it has a specific AP from owner coefficient that you'll have to figure out. Does the pet show up on WCL?

Seems to, I swung at a dummy for 20 minutes with logs on https://www.warcraftlogs.com/reports/wG9kWYcBL6QtF7Jg/#fight=last&type=summary&source=12&view=events

Khazakdk commented:
Taking off all gear except weapon and legendary, I have 527 strength, 26.9 wdps 18% mastery.

(527 + 26.9 * 6) * 1.5 * .97 * 1.18 = 1181.91

In game it hit for 1433.  The hits are extremely consistent too.  Towards the end of the log I linked there's 6 hits in a row for 2312.  Shouldn't the standard 5% deviation apply?
![image](https://user-images.githubusercontent.com/10967144/96387770-46fb1e00-116a-11eb-8c58-e9effa2af1a5.png)

Melekus commented:
I have not seen the 5% standard variance apply on any pet damage when I was testing on PTR, safe to say it's only enabled on players (whether it's a bug or not is another matter)

Melekus commented:
Looking at the log, the pets show a 1:1 ap from owner conversion.
Can you provide more datapoints with different strength/wdps/versatility/mastery values?

Khazakdk commented:
(str + wps * 6) * 1.5 * .97 * mastery * vers

(527 + 26.9 * 6) * 1.5 * .97 * 1.18 = 1181.91
actual 1433
(527 + 26.9 * 6) * 1.5 * .97 * 1.18 * x = 1433
x = 1.21244

(865 + 26.9 * 6) * 1.5 * .97 * 1.31 * 1.04 = 2034
actual 2312
(865 + 26.9 * 6) * 1.5 * .97 * 1.31 * 1.04 * x = 2312
x = 1.13633

(500 + 6.8 * 6) * 1.5 * .97 * 1.18 = 928.4995
actual 1182
(500 + 6.8 * 6) * 1.5 * .97 * 1.18 * x = 1182
x = 1.27

(568 + 6.8 * 6) * 1.5 * .97 * 1.21 = 1071.8228
actual 1329
(500 + 6.8 * 6) * 1.5 * .97 * 1.21 * x = 1329
x = 1.39585

(568 + 6.8 * 6) * 1.5 * .97 * 1.24 * 1.02 = 1120.3649
actual 1398
(568 + 6.8 * 6) * 1.5 * .97 * 1.24 * 1.02 * x = 1398
x = 1.24781

(604 + 6.8 * 6) * 1.5 * .97 * 1.27 * 1.02 = 1215.3236
actual 1496
(604 + 6.8 * 6) * 1.5 * .97 * 1.27 * 1.02 * x = 1496
x = 1.12095

(631 + 26.9 * 6) * 1.5 * .97 * 1.27 * 1.02 = 1493.521
actual 1770
(631 + 26.9 * 6) * 1.5 * .97 * 1.27 * 1.02 * x = 1770
x = 1.18512


Started with no gear except weapon and legendary, then start adding trinkets, rings, a helm.  Swapped between a ilvl 138 weapon and ilvl 36.

How did you determine the pets ap from the logs?  That's a handy trick for the future.

~My math skills aren't the best, but I think we're solving for an extra unknown multiplier?  (str + wps * 6) * 1.5 * .97 * mastery * vers * X = damage~

~So the next step would be taking the above data points, setting them equal to the actual damage and solving for X?~

Melekus commented:
Select an actor (player, pet, guardian), go to the Resources Tab and select Attack Power.

https://www.warcraftlogs.com/reports/wG9kWYcBL6QtF7Jg/#fight=last&type=resources&source=43&view=events&spell=1001

Khazakdk commented:
After those tests I did some better ones, testing one stat change at a time and trying to predict what would happen.

(631 + 26.9 * 6) * 1.5 * .97 * 1.27 * 1.02 = 1493.521
actual 1770
hypothesis: dropping to .85% vers will decrease dps by 1.15%
(631 + 26.9 * 6) * 1.5 * .97 * 1.27 * 1.0085 = 1476.6823 (1.12745% decrease)
if actual was 1770, new actual will be 1750.044135 (1770 * .9887255)
new actual was 1750


(597 + 26.9 * 6) * 1.5 * .97 * 1.206 = 1330.7872
actual 1587
hypothesis: adding 6% mastery will increase dps by 6%
(597 + 26.9 * 6) * 1.5 * .97 * 1.27 = 1401.40944 (5.3068% increase)
if actual was 1587, new actual will be 1671.218916
new actual was 1671


(597 + 26.9 * 6) * 1.5 * .97 * 1.206 = 1330.7872
actual 1587
hypothesis: adding 77 strength will increase dps by 13%?  (not as good at eyeballing what a numerical change will do)
(674 + 26.9 * 6) * 1.5 * .97 * 1.206 = 1465.90 (10.1528 increase)
if actual was 1587, new actual will be 1748.124936
new actual was 1722

Maybe there's something extra going on with strength?

Melekus commented:
I made a sheet to compare your last results, it looks like there's some sort of base damage *somewhere*.

https://docs.google.com/spreadsheets/d/1TGK_EGaJRWWR3HVa-ETuLRCsI2cUo2Pf2ZA_Nme9r-k/edit#gid=2075804396

Khazakdk commented:
Given Chewie's bug report I'll add the extra base damage inside a if (bugs) condition and make a note to follow up it as we get closer to launch, whenever that is.

Khazakdk committed:
[Death Knight]: Finish implementation of Reanimated Shambler

Khazakdk commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -2666,6 +2674,122 @@ struct magus_pet_t : public death_knight_pet_t
+      p() -> o() -> trigger_festering_wound( state, 1, p() -> o() -> procs.fw_necroblast );
+    }
+
+    double calculate_direct_amount( action_state_t* state ) const override
+    {
+      if ( p() -> o() -> bugs )
+      {
+        // See https://github.com/SimCMinMax/WoW-BugTracker/issues/725
+        // we think the total ap for necroblast is (total_ap + base_amount) * ap_coefficient
+        // action_t::calculate_direct_amount gets the total ap with base_amount + ap_coefficient * total_ap
+        // because of this, I need to directly modify the ap for the action instead of setting the base damage in
+        // the necroblast constructor as the latter method throws off the calcuations and doesn't reflect in game values
+        state->attack_power += 150;
+      }
+
+      double da = pet_action_t::calculate_direct_amount( state );
+
+      return da;
+    }
+  };
```
My gut says reaching out and modifying the state like this is bad but I'm not sure how to get around it.  I'll take any suggestions on how to do this better.  Based on the math from the spreadsheet, the extra base damage seems to be added to the AP amount before multiplying by the coefficient.  So I need to modify the AP of just this action.  I validated that only the damage numbers from necroblast are affected by this.  At this point just hoping Blizzard fixes it.

Melekus replied:
Yeah as long as it's reasonably close to the damage seen ingame and it doesn't negatively affect the rest of the simulation it's fine.

Melekus commented on engine/class_modules/sc_death_knight.cpp:
```diff
@@ -3149,6 +3273,11 @@ struct melee_t : public death_knight_melee_attack_t
         trigger_bloodworm();
       }
+
+      if ( p() -> legendary.reanimated_shambler -> ok() )
```
Does it only trigger on melee auto attacks?

Khazakdk replied:
Oh duh I need to move it up to death_knight_melee_attack_t.  Forgot bloodworms was autos only and thought melee_t was a part of every ability that involve a melee hit.  Abilities can proc it too, based on the spelldata and what I'm seeing in game.  

`Proc Flags       : ..x.x... ........ ........ ........ 
                 : White Melee, Yellow Melee`

Khazakdk replied:
Proc numbers moved a bit after the change

Before
`Reanimated Shambler | 8.2 | 3.0 | 20.0 | 32.9s | 3.4s | 133.7s`

After
`Reanimated Shambler | 8.2 | 2.0 | 22.0 | 32.5s | 0.1s | 130.0s`

Khazakdk replied:
Proc numbers moved a bit after the change

Before
Reanimated Shambler | 8.2 | 3.0 | 20.0 | 32.9s | 3.4s | 133.7s

After
Reanimated Shambler | 8.2 | 2.0 | 17.0 | 33.0s | 0.1s | 135.6s

I'm not 100% on the accuracy of putting it in execute as it would probably be more technically correct to make sure the attack isn't parried.  But impact gets run multiple times for AoE so shouldn't have it there either.  Unless each hit of SS cleave has a chance to proc it...

Melekus replied:
It should be implemented like a regular proc : simc handles this by registering special effects and callbacks with specific proc flags and actions/buffs to execute.

Khazakdk committed:
[Death Knight]: Moved shambler to proc on all melee abilities

Melekus committed:
[Death Knight] Update Reanimated Shambler proc system

Replace the death_knight_melee_attack_t rppm trigger attempt by a special effect / callback system based on rppm and proc flags data.
Remove unneeded rppm.reanimated_shambler object, disable the check for spec.festering_wound on festering_wound debuff application, as Necroblast seems to apply festering wounds even for non-unholy specs.
Register the special effect on detection of spellID 334836, associated with spell ID 6949 in item_runeforge.inc
